# Graffiti

In [1]:
import pandas as pd 
import numpy as np

In [2]:
df = pd.read_csv('../../data/raw/Graffiti.csv')

//anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3057: DtypeWarning: Columns (19) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
df.head()

CaseID                  Opened                  Closed  \
0  11767872  12/04/2019 07:59:00 AM  12/04/2019 10:15:12 PM   
1  11769150  12/04/2019 11:12:00 AM                     NaN   
2  11768407  12/04/2019 09:20:53 AM                     NaN   
3  11767877  12/04/2019 07:59:56 AM  12/04/2019 10:15:11 PM   
4  11771899  12/04/2019 08:38:00 PM                     NaN   

                  Updated  Status  \
0  12/04/2019 10:15:12 PM  Closed   
1  12/04/2019 11:30:41 AM    Open   
2  12/04/2019 09:20:56 AM    Open   
3  12/04/2019 10:15:11 PM  Closed   
4  12/04/2019 09:10:05 PM    Open   

                                        Status Notes  \
0  Request is a duplicate and has been previously...   
1                                           accepted   
2                                               Open   
3  Request is a duplicate and has been previously...   
4                                           accepted   

              Responsible Agency  Category  \
0                  DPW Ops Queue  Graffiti   
1                  DPW Ops Queue  Graffiti   
2  Clear Channel - Transit Queue  Graffiti   
3                  DPW Ops Queue  Graffiti   
4                  DPW Ops Queue  Graffiti   

                                       Request Type  \
0  Graffiti on Other_enter_additional_details_below   
1                            Graffiti on Signal_box   
2              Graffiti on Transit_Shelter_Platform   
3                                Graffiti on Street   
4                                  Graffiti on Pole   

                                    Request Details  \
0  Other_enter_additional_details_below - Offensive   
1                        Signal_box - Not_Offensive   
2          Transit_Shelter_Platform - Not_Offensive   
3                                Street - Offensive   
4                              Pole - Not_Offensive   

                                        Address         Street  \
0        Intersection of 22ND ST and MISSION ST        22ND ST   
1  Intersection of SAN BRUNO AVE and SILVER AVE  SAN BRUNO AVE   
2  1170 MCALLISTER ST, SAN FRANCISCO, CA, 94115  MCALLISTER ST   
3      Intersection of 11TH AVE and LINCOLN WAY       11TH AVE   
4    1198 PACIFIC AVE, SAN FRANCISCO, CA, 94133    PACIFIC AVE   

   Supervisor District      Neighborhood Police District   Latitude  \
0                  9.0           Mission         MISSION  37.755459   
1                  9.0           Portola         BAYVIEW  37.732437   
2                  5.0  Western Addition        NORTHERN  37.778758   
3                  5.0      Inner Sunset         TARAVAL  37.765886   
4                  3.0      Russian Hill         CENTRAL  37.795982   

    Longitude                         Point          Source  \
0 -122.418761  (37.75545911, -122.41876097)  Mobile/Open311   
1 -122.405609  (37.73243713, -122.40560913)           Phone   
2 -122.431420  (37.77875764, -122.43142005)  Mobile/Open311   
3 -122.468599    (37.7658857, -122.4685986)             Web   
4 -122.414897    (37.7959824, -122.4148973)             Web   

                                           Media URL  
0  http://mobile311.sfgov.org/reports/11767872/ph...  
1                                                NaN  
2  http://mobile311.sfgov.org/reports/11768407/ph...  
3                                                NaN  
4                                                NaN

In [4]:
df.dtypes

CaseID                   int64
Opened                  object
Closed                  object
Updated                 object
Status                  object
Status Notes            object
Responsible Agency      object
Category                object
Request Type            object
Request Details         object
Address                 object
Street                  object
Supervisor District    float64
Neighborhood            object
Police District         object
Latitude               float64
Longitude              float64
Point                   object
Source                  object
Media URL               object
dtype: object

## Rename columns

In [5]:
df.rename(columns={'Latitude': 'Y',
                  'Longitude': 'X'}, inplace=True)
df.head()

CaseID                  Opened                  Closed  \
0  11767872  12/04/2019 07:59:00 AM  12/04/2019 10:15:12 PM   
1  11769150  12/04/2019 11:12:00 AM                     NaN   
2  11768407  12/04/2019 09:20:53 AM                     NaN   
3  11767877  12/04/2019 07:59:56 AM  12/04/2019 10:15:11 PM   
4  11771899  12/04/2019 08:38:00 PM                     NaN   

                  Updated  Status  \
0  12/04/2019 10:15:12 PM  Closed   
1  12/04/2019 11:30:41 AM    Open   
2  12/04/2019 09:20:56 AM    Open   
3  12/04/2019 10:15:11 PM  Closed   
4  12/04/2019 09:10:05 PM    Open   

                                        Status Notes  \
0  Request is a duplicate and has been previously...   
1                                           accepted   
2                                               Open   
3  Request is a duplicate and has been previously...   
4                                           accepted   

              Responsible Agency  Category  \
0                  DPW Ops Queue  Graffiti   
1                  DPW Ops Queue  Graffiti   
2  Clear Channel - Transit Queue  Graffiti   
3                  DPW Ops Queue  Graffiti   
4                  DPW Ops Queue  Graffiti   

                                       Request Type  \
0  Graffiti on Other_enter_additional_details_below   
1                            Graffiti on Signal_box   
2              Graffiti on Transit_Shelter_Platform   
3                                Graffiti on Street   
4                                  Graffiti on Pole   

                                    Request Details  \
0  Other_enter_additional_details_below - Offensive   
1                        Signal_box - Not_Offensive   
2          Transit_Shelter_Platform - Not_Offensive   
3                                Street - Offensive   
4                              Pole - Not_Offensive   

                                        Address         Street  \
0        Intersection of 22ND ST and MISSION ST        22ND ST   
1  Intersection of SAN BRUNO AVE and SILVER AVE  SAN BRUNO AVE   
2  1170 MCALLISTER ST, SAN FRANCISCO, CA, 94115  MCALLISTER ST   
3      Intersection of 11TH AVE and LINCOLN WAY       11TH AVE   
4    1198 PACIFIC AVE, SAN FRANCISCO, CA, 94133    PACIFIC AVE   

   Supervisor District      Neighborhood Police District          Y  \
0                  9.0           Mission         MISSION  37.755459   
1                  9.0           Portola         BAYVIEW  37.732437   
2                  5.0  Western Addition        NORTHERN  37.778758   
3                  5.0      Inner Sunset         TARAVAL  37.765886   
4                  3.0      Russian Hill         CENTRAL  37.795982   

            X                         Point          Source  \
0 -122.418761  (37.75545911, -122.41876097)  Mobile/Open311   
1 -122.405609  (37.73243713, -122.40560913)           Phone   
2 -122.431420  (37.77875764, -122.43142005)  Mobile/Open311   
3 -122.468599    (37.7658857, -122.4685986)             Web   
4 -122.414897    (37.7959824, -122.4148973)             Web   

                                           Media URL  
0  http://mobile311.sfgov.org/reports/11767872/ph...  
1                                                NaN  
2  http://mobile311.sfgov.org/reports/11768407/ph...  
3                                                NaN  
4                                                NaN

## Empty values

In [6]:
for col in df:
    print(col, df[col].isnull().sum() )

CaseID 0
Opened 0
Closed 20905
Updated 0
Status 0
Status Notes 184027
Responsible Agency 0
Category 0
Request Type 0
Request Details 22
Address 0
Street 6967
Supervisor District 6967
Neighborhood 7555
Police District 7725
Y 0
X 0
Point 0
Source 0
Media URL 299726


Closed nan values are fine, because the case is still open and we will only look at the open date. Status notes and request type are also not useful. Basically we need to take care of nan values for Location, Supervisor District, Neighborhood and Police District

In [7]:
df[df['Supervisor District'].isnull()]

CaseID                  Opened                  Closed  \
479     11760592  12/02/2019 04:05:00 PM                     NaN   
1802    11699075  11/17/2019 08:21:07 AM  11/27/2019 10:00:12 PM   
2113    11722224  11/22/2019 08:45:00 AM                     NaN   
2715    11649298  11/05/2019 12:52:00 PM  11/21/2019 09:00:57 AM   
2997    11559723  10/16/2019 03:21:00 PM                     NaN   
3092    11704015  11/18/2019 11:45:00 AM  11/19/2019 10:35:48 AM   
3341    11634897  11/02/2019 09:11:15 AM                     NaN   
3649    11696384  11/16/2019 10:24:00 AM  11/16/2019 12:21:45 PM   
3794    11679848  11/12/2019 02:54:00 PM  11/15/2019 10:31:08 AM   
6706    11619810  10/29/2019 06:15:00 PM  10/29/2019 10:15:35 PM   
6765    11613436  10/28/2019 02:20:00 PM  10/29/2019 10:47:09 AM   
6792    11607427  10/27/2019 08:17:00 AM  10/29/2019 02:23:33 PM   
7031    11594920  10/24/2019 09:48:00 AM  10/26/2019 01:46:19 PM   
7441    11575861  10/20/2019 11:20:00 AM  10/24/2019 10:08:00 AM   
7816    11579641  10/21/2019 10:50:00 AM  10/21/2019 11:51:55 AM   
8041    11567576  10/18/2019 09:41:00 AM  10/19/2019 09:53:02 AM   
8722    11539133  10/11/2019 05:40:00 PM  10/12/2019 08:13:20 AM   
9751    11498250  10/03/2019 08:09:00 AM  10/03/2019 01:15:38 PM   
10448   11451047  09/22/2019 04:45:00 PM  09/29/2019 09:16:35 AM   
11040    5699352  03/25/2016 11:25:05 AM  04/02/2016 09:45:00 AM   
11716    9897429  12/06/2018 12:22:57 AM                     NaN   
11801    4276393  12/06/2014 01:06:30 PM  12/24/2014 07:48:01 AM   
12089    7520874  08/18/2017 11:11:30 AM                     NaN   
12144    7515509  08/17/2017 11:13:20 AM                     NaN   
12153    4242950  11/25/2014 12:06:13 PM  12/03/2014 09:00:00 AM   
12279    4234171  11/22/2014 04:01:56 PM  12/02/2014 12:00:00 PM   
13370    5671708  03/16/2016 02:10:25 PM  03/22/2016 12:21:00 PM   
13384    5671531  03/16/2016 01:38:53 PM  03/22/2016 12:21:00 PM   
13751    6762812  01/23/2017 04:13:10 PM  12/13/2018 10:30:25 PM   
13825    6398019  10/08/2016 04:04:54 PM  12/06/2018 10:30:57 PM   
...          ...                     ...                     ...   
479106   8431690  12/27/2017 11:12:00 AM                     NaN   
479240   8452156  01/02/2018 09:13:00 AM                     NaN   
479517   8311336  11/26/2017 02:28:00 PM                     NaN   
480079   8641686  02/16/2018 09:57:00 AM  02/16/2018 11:38:00 AM   
480814   8536296  01/22/2018 10:15:00 AM                     NaN   
481339   8328127  11/30/2017 09:29:00 AM                     NaN   
482604   8451498  01/02/2018 06:38:00 AM  01/08/2018 05:00:37 PM   
482679   8405831  12/19/2017 02:14:00 PM                     NaN   
482829   8701572  03/03/2018 10:20:00 AM  03/03/2018 01:39:00 PM   
482866   8587598  02/03/2018 08:01:00 AM  02/03/2018 10:25:37 AM   
488257   8960949  05/05/2018 03:22:00 AM  01/30/2019 11:00:20 PM   
490960   8849710  04/09/2018 03:29:00 PM  04/09/2018 06:25:00 PM   
491542   8976757  05/08/2018 08:58:00 PM  05/08/2018 09:10:43 PM   
498428   8759205  03/18/2018 10:07:00 PM  03/18/2018 10:46:00 PM   
499998   8885426  04/18/2018 06:54:00 AM  06/04/2018 01:30:23 PM   
503187  10612655  03/17/2019 06:08:00 PM  03/20/2019 09:33:15 AM   
508082  11025025  06/20/2019 05:51:00 PM  06/20/2019 06:48:00 PM   
510754  11009751  06/17/2019 05:16:00 PM  06/17/2019 07:35:00 PM   
513375  11025040  06/20/2019 05:54:00 PM  06/21/2019 08:04:00 AM   
517391  10786233  04/27/2019 02:25:00 PM  04/27/2019 03:33:00 PM   
518861  11372586  09/05/2019 10:35:00 AM  09/06/2019 09:18:33 AM   
518869  11395526  09/10/2019 11:22:00 AM  09/17/2019 11:00:08 AM   
519142  11352919  09/01/2019 07:55:00 AM  09/02/2019 11:15:22 AM   
519392  11104786  07/09/2019 01:14:28 PM                     NaN   
520055  11328793  08/27/2019 08:04:00 AM  08/29/2019 11:04:46 AM   
520385  11381663  09/07/2019 10:04:00 AM  09/11/2019 11:34:24 AM   
523211  11390435  09/09/2019 12:22:00 PM  09/17/2019 08:51:39 AM  

We see a problem. The rows with no district are the ones with location (0,0). Since we are going to work based on locations. Drop all the locations with (0,0)

In [8]:
df = df[df.Point != '(0.0, 0.0)']
df[df['Supervisor District'].isnull()]

CaseID                  Opened                  Closed  \
479     11760592  12/02/2019 04:05:00 PM                     NaN   
1802    11699075  11/17/2019 08:21:07 AM  11/27/2019 10:00:12 PM   
2113    11722224  11/22/2019 08:45:00 AM                     NaN   
2715    11649298  11/05/2019 12:52:00 PM  11/21/2019 09:00:57 AM   
3092    11704015  11/18/2019 11:45:00 AM  11/19/2019 10:35:48 AM   
3649    11696384  11/16/2019 10:24:00 AM  11/16/2019 12:21:45 PM   
3794    11679848  11/12/2019 02:54:00 PM  11/15/2019 10:31:08 AM   
6706    11619810  10/29/2019 06:15:00 PM  10/29/2019 10:15:35 PM   
6765    11613436  10/28/2019 02:20:00 PM  10/29/2019 10:47:09 AM   
6792    11607427  10/27/2019 08:17:00 AM  10/29/2019 02:23:33 PM   
7031    11594920  10/24/2019 09:48:00 AM  10/26/2019 01:46:19 PM   
7441    11575861  10/20/2019 11:20:00 AM  10/24/2019 10:08:00 AM   
8041    11567576  10/18/2019 09:41:00 AM  10/19/2019 09:53:02 AM   
8722    11539133  10/11/2019 05:40:00 PM  10/12/2019 08:13:20 AM   
9751    11498250  10/03/2019 08:09:00 AM  10/03/2019 01:15:38 PM   
10448   11451047  09/22/2019 04:45:00 PM  09/29/2019 09:16:35 AM   
13751    6762812  01/23/2017 04:13:10 PM  12/13/2018 10:30:25 PM   
18030    8620570  02/11/2018 03:32:00 PM  02/11/2018 04:29:00 PM   
18136    8196221  10/27/2017 09:28:06 PM  10/27/2017 10:36:00 PM   
22016    4462650  02/05/2015 11:45:24 AM  02/17/2015 03:14:36 AM   
29829    9109284  06/08/2018 12:05:52 AM  12/07/2018 07:27:21 AM   
44660    9440107  08/22/2018 09:52:00 PM  08/24/2018 08:21:00 AM   
57331   10307300  01/02/2019 04:49:50 PM  01/02/2019 05:40:00 PM   
216287   3230556  01/03/2014 03:54:13 PM  01/15/2014 08:18:46 AM   
216828   3380405  02/18/2014 04:25:00 PM  03/07/2014 05:34:49 AM   
217914   3120793  12/02/2013 08:50:29 AM  12/05/2013 10:09:16 AM   
219871   3313962  01/28/2014 09:30:13 AM  01/30/2014 06:16:23 AM   
220496   3277452  01/16/2014 09:06:12 AM  02/03/2014 10:18:02 AM   
226352   3362146  02/12/2014 04:05:29 PM  03/07/2014 02:52:46 AM   
227905   3311634  01/27/2014 02:31:53 PM  01/27/2014 05:00:53 PM   
...          ...                     ...                     ...   
404334   6449171  10/22/2016 01:28:10 PM  12/31/2016 12:00:00 AM   
410259   6920938  03/09/2017 03:18:57 PM  03/14/2017 11:30:28 AM   
411020   7023647  04/07/2017 10:22:46 PM  04/07/2017 10:42:06 PM   
412548   6904548  03/05/2017 12:24:02 PM  04/30/2017 01:25:25 PM   
415806   6856755  02/18/2017 09:03:01 PM  02/22/2017 08:54:58 AM   
418851   7062584  04/19/2017 05:14:20 PM  04/20/2017 07:45:53 AM   
420285   6893267  03/01/2017 03:53:41 PM  03/05/2017 05:25:11 PM   
422338   7155867  05/16/2017 02:26:20 PM  06/15/2017 07:25:20 PM   
422709   7040143  04/12/2017 11:05:16 PM  04/13/2017 02:22:47 AM   
423608   6841241  02/14/2017 11:11:31 AM  12/20/2018 10:30:29 PM   
427365   7047005  04/14/2017 07:00:37 PM  04/23/2017 12:27:56 AM   
428396   7145080  05/13/2017 10:42:18 AM  05/17/2017 12:25:39 PM   
433008   7444378  07/31/2017 04:16:55 PM  08/05/2017 10:25:49 AM   
436803   7218427  06/02/2017 10:57:43 PM  06/09/2017 10:30:40 PM   
438088   7377491  07/15/2017 11:51:14 AM  07/17/2017 08:20:13 AM   
438977   7341735  07/06/2017 12:45:18 PM  07/10/2017 02:31:59 PM   
441542   7371166  07/13/2017 04:35:33 PM  07/14/2017 06:18:47 AM   
442981   7197869  05/28/2017 02:56:31 PM  06/03/2017 09:26:15 AM   
465701   8561552  01/28/2018 09:54:26 AM  01/28/2018 10:04:00 AM   
476852   8645780  02/17/2018 11:11:00 AM  02/26/2018 05:00:10 PM   
480079   8641686  02/16/2018 09:57:00 AM  02/16/2018 11:38:00 AM   
482604   8451498  01/02/2018 06:38:00 AM  01/08/2018 05:00:37 PM   
491542   8976757  05/08/2018 08:58:00 PM  05/08/2018 09:10:43 PM   
498428   8759205  03/18/2018 10:07:00 PM  03/18/2018 10:46:00 PM   
518869  11395526  09/10/2019 11:22:00 AM  09/17/2019 11:00:08 AM   
519142  11352919  09/01/2019 07:55:00 AM  09/02/2019 11:15:22 AM   
520055  11328793  08/27/2019 08:04:00 AM  08/29/2019 11:04:46 AM  

We got rid from a lot of missing location rows, but we are still missing some supervisor districts. We can get the correct district using the location later with our class.

In [9]:
print('Nan on Neighborhood', df[df['Neighborhood'].isnull()].shape[0])
print('Nan on Police District', df[df['Police District'].isnull()].shape[0])

Nan on Neighborhood 510
Nan on Police District 680


In [10]:
df[df['Neighborhood'].isnull()]

CaseID                  Opened                  Closed  \
479     11760592  12/02/2019 04:05:00 PM                     NaN   
1802    11699075  11/17/2019 08:21:07 AM  11/27/2019 10:00:12 PM   
2113    11722224  11/22/2019 08:45:00 AM                     NaN   
2715    11649298  11/05/2019 12:52:00 PM  11/21/2019 09:00:57 AM   
3092    11704015  11/18/2019 11:45:00 AM  11/19/2019 10:35:48 AM   
3649    11696384  11/16/2019 10:24:00 AM  11/16/2019 12:21:45 PM   
3794    11679848  11/12/2019 02:54:00 PM  11/15/2019 10:31:08 AM   
6706    11619810  10/29/2019 06:15:00 PM  10/29/2019 10:15:35 PM   
6765    11613436  10/28/2019 02:20:00 PM  10/29/2019 10:47:09 AM   
6792    11607427  10/27/2019 08:17:00 AM  10/29/2019 02:23:33 PM   
7031    11594920  10/24/2019 09:48:00 AM  10/26/2019 01:46:19 PM   
7441    11575861  10/20/2019 11:20:00 AM  10/24/2019 10:08:00 AM   
8041    11567576  10/18/2019 09:41:00 AM  10/19/2019 09:53:02 AM   
8722    11539133  10/11/2019 05:40:00 PM  10/12/2019 08:13:20 AM   
9751    11498250  10/03/2019 08:09:00 AM  10/03/2019 01:15:38 PM   
10448   11451047  09/22/2019 04:45:00 PM  09/29/2019 09:16:35 AM   
13751    6762812  01/23/2017 04:13:10 PM  12/13/2018 10:30:25 PM   
18030    8620570  02/11/2018 03:32:00 PM  02/11/2018 04:29:00 PM   
18136    8196221  10/27/2017 09:28:06 PM  10/27/2017 10:36:00 PM   
18913    6906975  03/06/2017 10:07:39 AM  03/06/2017 12:10:14 PM   
19480    4832264  06/16/2015 03:03:28 PM  06/30/2015 08:05:43 AM   
22016    4462650  02/05/2015 11:45:24 AM  02/17/2015 03:14:36 AM   
29829    9109284  06/08/2018 12:05:52 AM  12/07/2018 07:27:21 AM   
30634    9338187  07/30/2018 09:36:00 PM                     NaN   
38152    8525438  01/19/2018 07:28:00 AM  01/19/2018 09:31:00 AM   
38545    6515195  11/09/2016 04:11:00 PM  01/16/2018 04:35:00 PM   
41320    8359246  12/07/2017 02:49:00 PM  08/23/2018 11:00:08 AM   
41485    8417838  12/22/2017 12:37:00 PM  08/23/2018 11:00:06 AM   
41755    6906983  03/06/2017 10:09:19 AM  09/14/2018 03:31:09 PM   
44660    9440107  08/22/2018 09:52:00 PM  08/24/2018 08:21:00 AM   
...          ...                     ...                     ...   
466062   8505621  01/14/2018 08:50:00 PM  01/25/2018 06:22:00 AM   
466492   8527883  01/19/2018 02:26:00 PM  01/21/2018 11:49:40 AM   
467052   8608232  02/08/2018 10:02:00 AM  02/09/2018 09:43:58 AM   
467546   8320292  11/28/2017 01:49:00 PM  12/10/2017 11:25:41 AM   
467948   8503988  01/14/2018 11:33:00 AM  01/25/2018 06:21:00 AM   
469005   8481237  01/09/2018 08:01:00 AM  01/18/2018 10:49:18 AM   
469777   8488316  01/10/2018 12:29:00 PM  01/12/2018 10:25:18 AM   
469848   8539127  01/22/2018 04:25:00 PM  01/23/2018 07:41:00 AM   
470470   8362705  12/08/2017 11:28:00 AM  12/08/2017 02:41:00 PM   
471287   8457612  01/03/2018 09:37:00 AM  01/06/2018 09:15:23 AM   
472864   8588619  02/03/2018 12:14:00 PM  02/05/2018 10:28:59 AM   
475946   8554725  01/26/2018 09:50:00 AM  01/26/2018 12:22:00 PM   
476573   8505369  01/14/2018 06:42:00 PM  01/17/2018 11:10:09 AM   
476852   8645780  02/17/2018 11:11:00 AM  02/26/2018 05:00:10 PM   
477181   8653242  02/19/2018 09:12:00 PM  09/14/2018 11:00:57 AM   
480079   8641686  02/16/2018 09:57:00 AM  02/16/2018 11:38:00 AM   
480618   8444040  12/30/2017 08:03:00 AM  12/30/2017 10:14:14 AM   
481128   8488293  01/10/2018 12:26:00 PM  01/12/2018 10:28:15 AM   
482008   8440715  12/29/2017 10:22:00 AM  12/30/2017 09:44:03 AM   
482604   8451498  01/02/2018 06:38:00 AM  01/08/2018 05:00:37 PM   
482968   8493836  01/11/2018 02:37:00 PM  01/29/2018 10:30:07 AM   
491542   8976757  05/08/2018 08:58:00 PM  05/08/2018 09:10:43 PM   
493699   8864529  04/12/2018 03:37:00 PM  04/14/2018 09:43:11 AM   
498428   8759205  03/18/2018 10:07:00 PM  03/18/2018 10:46:00 PM   
518869  11395526  09/10/2019 11:22:00 AM  09/17/2019 11:00:08 AM   
519142  11352919  09/01/2019 07:55:00 AM  09/02/2019 11:15:22 AM   
520055  11328793  08/27/2019 08:04:00 AM  08/29/2019 11:04:46 AM  

There are less than 510 rows with nan out of 521203, that is 0.09%. Therefore, we will drop them without risk of losing data

In [11]:
df.dropna(subset=['Neighborhood'], inplace=True)
df.dropna(subset=['Police District'], inplace=True)
df.dropna(subset=['Supervisor District'], inplace=True)

In [12]:
print('Nan on Neighborhood', df[df['Neighborhood'].isnull()].shape[0])
print('Nan on Police District', df[df['Police District'].isnull()].shape[0])
print('Nan on Supervisor District', df[df['Supervisor District'].isnull()].shape[0])

Nan on Neighborhood 0
Nan on Police District 0
Nan on Supervisor District 0


## Datetime objects

In [13]:
df['Opened'] = pd.to_datetime(df['Opened'] )
df['Closed'] = pd.to_datetime(df['Closed'] )
df['Updated'] = pd.to_datetime(df['Updated'] )

In [14]:
df['Year'] = df['Opened'].dt.year
df['Month'] = df['Opened'].dt.month
df['Day'] = df['Opened'].dt.day

In [15]:
df.head()

CaseID              Opened              Closed             Updated  \
0  11767872 2019-12-04 07:59:00 2019-12-04 22:15:12 2019-12-04 22:15:12   
1  11769150 2019-12-04 11:12:00                 NaT 2019-12-04 11:30:41   
2  11768407 2019-12-04 09:20:53                 NaT 2019-12-04 09:20:56   
3  11767877 2019-12-04 07:59:56 2019-12-04 22:15:11 2019-12-04 22:15:11   
4  11771899 2019-12-04 20:38:00                 NaT 2019-12-04 21:10:05   

   Status                                       Status Notes  \
0  Closed  Request is a duplicate and has been previously...   
1    Open                                           accepted   
2    Open                                               Open   
3  Closed  Request is a duplicate and has been previously...   
4    Open                                           accepted   

              Responsible Agency  Category  \
0                  DPW Ops Queue  Graffiti   
1                  DPW Ops Queue  Graffiti   
2  Clear Channel - Transit Queue  Graffiti   
3                  DPW Ops Queue  Graffiti   
4                  DPW Ops Queue  Graffiti   

                                       Request Type  \
0  Graffiti on Other_enter_additional_details_below   
1                            Graffiti on Signal_box   
2              Graffiti on Transit_Shelter_Platform   
3                                Graffiti on Street   
4                                  Graffiti on Pole   

                                    Request Details  ...      Neighborhood  \
0  Other_enter_additional_details_below - Offensive  ...           Mission   
1                        Signal_box - Not_Offensive  ...           Portola   
2          Transit_Shelter_Platform - Not_Offensive  ...  Western Addition   
3                                Street - Offensive  ...      Inner Sunset   
4                              Pole - Not_Offensive  ...      Russian Hill   

  Police District          Y           X                         Point  \
0         MISSION  37.755459 -122.418761  (37.75545911, -122.41876097)   
1         BAYVIEW  37.732437 -122.405609  (37.73243713, -122.40560913)   
2        NORTHERN  37.778758 -122.431420  (37.77875764, -122.43142005)   
3         TARAVAL  37.765886 -122.468599    (37.7658857, -122.4685986)   
4         CENTRAL  37.795982 -122.414897    (37.7959824, -122.4148973)   

           Source                                          Media URL  Year  \
0  Mobile/Open311  http://mobile311.sfgov.org/reports/11767872/ph...  2019   
1           Phone                                                NaN  2019   
2  Mobile/Open311  http://mobile311.sfgov.org/reports/11768407/ph...  2019   
3             Web                                                NaN  2019   
4             Web                                                NaN  2019   

  Month Day  
0    12   4  
1    12   4  
2    12   4  
3    12   4  
4    12   4  

[5 rows x 23 columns]

## Readability

In [16]:
df['Police District'] = df['Police District'].astype(str).str.lower() 
df['Supervisor District'] = df['Supervisor District'].astype(str).str.lower() 
df['Neighborhood'] = df['Neighborhood'].astype(str).str.lower() 
df['Address'] = df['Address'].astype(str).str.lower() 
df.head()

CaseID              Opened              Closed             Updated  \
0  11767872 2019-12-04 07:59:00 2019-12-04 22:15:12 2019-12-04 22:15:12   
1  11769150 2019-12-04 11:12:00                 NaT 2019-12-04 11:30:41   
2  11768407 2019-12-04 09:20:53                 NaT 2019-12-04 09:20:56   
3  11767877 2019-12-04 07:59:56 2019-12-04 22:15:11 2019-12-04 22:15:11   
4  11771899 2019-12-04 20:38:00                 NaT 2019-12-04 21:10:05   

   Status                                       Status Notes  \
0  Closed  Request is a duplicate and has been previously...   
1    Open                                           accepted   
2    Open                                               Open   
3  Closed  Request is a duplicate and has been previously...   
4    Open                                           accepted   

              Responsible Agency  Category  \
0                  DPW Ops Queue  Graffiti   
1                  DPW Ops Queue  Graffiti   
2  Clear Channel - Transit Queue  Graffiti   
3                  DPW Ops Queue  Graffiti   
4                  DPW Ops Queue  Graffiti   

                                       Request Type  \
0  Graffiti on Other_enter_additional_details_below   
1                            Graffiti on Signal_box   
2              Graffiti on Transit_Shelter_Platform   
3                                Graffiti on Street   
4                                  Graffiti on Pole   

                                    Request Details  ...      Neighborhood  \
0  Other_enter_additional_details_below - Offensive  ...           mission   
1                        Signal_box - Not_Offensive  ...           portola   
2          Transit_Shelter_Platform - Not_Offensive  ...  western addition   
3                                Street - Offensive  ...      inner sunset   
4                              Pole - Not_Offensive  ...      russian hill   

  Police District          Y           X                         Point  \
0         mission  37.755459 -122.418761  (37.75545911, -122.41876097)   
1         bayview  37.732437 -122.405609  (37.73243713, -122.40560913)   
2        northern  37.778758 -122.431420  (37.77875764, -122.43142005)   
3         taraval  37.765886 -122.468599    (37.7658857, -122.4685986)   
4         central  37.795982 -122.414897    (37.7959824, -122.4148973)   

           Source                                          Media URL  Year  \
0  Mobile/Open311  http://mobile311.sfgov.org/reports/11767872/ph...  2019   
1           Phone                                                NaN  2019   
2  Mobile/Open311  http://mobile311.sfgov.org/reports/11768407/ph...  2019   
3             Web                                                NaN  2019   
4             Web                                                NaN  2019   

  Month Day  
0    12   4  
1    12   4  
2    12   4  
3    12   4  
4    12   4  

[5 rows x 23 columns]

In [17]:
print("Min date: ",min(df['Opened']))
print("Max date: ",max(df['Opened']))

Min date:  2008-07-01 05:54:19
Max date:  2019-12-04 23:34:36


In [20]:
df.to_csv('../../data/processed/graffiti.csv')